File: leetcode/2570_merge _two_2d_arrays_by_summing_values.py
class Solution:
    def mergeArrays(
        self, nums1: List[List[int]], nums2: List[List[int]]
    ) -> List[List[int]]:
        array_map = {}

        # merge both arrays together and then add them to the map
        for key, val in nums1 + nums2:
            array_map[key] = array_map.get(key, 0) + val

        # sort the array and convert the array_map to an array
        return sorted([key, val] for key, val in array_map.items())

File: leetcode/2570_merge_two_2d_arrays_by_summing_values.js
/**
 * @param {number[][]} nums1
 * @param {number[][]} nums2
 * @return {number[][]}
 */
var mergeArrays = function (nums1, nums2) {
  let arrayMap = {};

  for (let [id, val] of nums1) {
    if (!arrayMap[id]) {
      arrayMap[id] = val;
    } else {
      arrayMap[id] = arrayMap[id] + val;
    }
  }

  for (let [id, val] of nums2) {
    if (!arrayMap[id]) {
      arrayMap[id] = val;
    } else {
      arrayMap[id] = arrayMap[id] + val;
    }
  }

  // Convert the arrayMap into a sorted array
  // The object should already be sorted
  return Object.entries(arrayMap).map(([key, val]) => [Number(key), val]);
};

// Time: O(N + M)
// Space: O(N + M)